src/python/pants/util/objects.py

```diff
@@ -393,32 +415,42 @@ def satisfied_by_type(self, obj_type):
     return issubclass(obj_type, self._types)
 
 
-class Collection(object):
-  """Constructs classes representing collections of objects of a particular type.
+class TypedCollection(TypeConstraint):
+  """A `TypeConstraint` which accepts a BasicTypeConstraint and validates a collection."""
 
-  The produced class will expose its values under a field named dependencies - this is a stable API
-  which may be consumed e.g. over FFI from the engine.
+  @classmethod
+  def _generate_variance_symbol(cls, constraint):
+    return '[{}]'.format(constraint._variance_symbol)
```

Comment:
I suspect this would be much clearer if it outputted something like:
`TypedCollection.of(Subclasses.of(Foo or Bar))` rather than `[+](Foo, Bar)`. It's non-obvious to me what the latter means.

Reply:
Sure, but that's kind of going away from the standard of extreme conciseness set by the existing uses of `variance_symbol` in the `__str__()` impls here. I will make a followup issue to cover how best to represent the `__str__()` of these. I like the very concise description, but it's completely incomprehensible / unsearchable to a newcomer. I think we may want to just remove the "variance symbol" idea altogether.

Reply:
After reading your other comment and thinking it may be a way to delete some code, I will try to address the `str` of the `TypeConstraint`s in this PR as well and break it out if necessary.

Reply:
The only difference now between `str` and `repr` is that `str` will use ` or ` between types, while `repr` will just use commas. I'm thinking I might make the error messages start using `str` instead of `repr` for this reason.